Modules/grpmodule.c

```diff
@@ -119,16 +120,47 @@ grp_getgrgid_impl(PyObject *module, PyObject *id)
         }
         Py_DECREF(py_int_id);
     }
+#ifdef HAVE_GETGRGID_R
+    Py_BEGIN_ALLOW_THREADS
+    int status, bufsize = NSS_BUFLEN_GROUP;
+    struct group grp;
 
-    if ((p = getgrgid(gid)) == NULL) {
+    buf = PyMem_RawMalloc(bufsize);
+
+    do {
+        status = getgrgid_r(gid, &grp, buf, bufsize, &p);
+        if (p != NULL || status != ERANGE) {
+            break;
+        }
+        if ((bufsize << 1) > (NSS_BUFLEN_GROUP << 10)) {
```

Comment:
This limit looks arbitrary. Use `if (bufsize > PY_SSIZE_T_MAX)` and set MemoryError by `PyErr_NoMemory()`.